src/kv/key.rs

```diff
@@ -33,15 +31,15 @@ impl ToKey for str {
 }
 
 /// A key in a structured key-value pair.
-#[derive(Clone)]
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
```

Comment:
I think one of the reasons we originally wrote these impls out by-hand was because there was some idea of adding an optional index to the key type that could be used as an optimization for lookup, but not considered when comparing.

That's not exactly obvious from the original code though, so maybe we can drop a comment there to suggest it.

```suggestion
// These impls must only be based on the as_str() representation of the key
// If a new field (such as an optional index) is added to the key they must not affect comparison
#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
```